### PR TITLE
[SPARK-4081] [mllib]  VectorIndexer

### PR DESCRIPTION
**Pull request by jkbradley**
**Ready for review!**

Since the original PR, I moved the code to the spark.ml API and renamed this to VectorIndexer.

This introduces a VectorIndexer class which does the following:
- VectorIndexer.fit(): collect statistics about how many values each feature in a dataset (RDD[Vector]) can take (limited by maxCategories)
  - Feature which exceed maxCategories are declared continuous, and the Model will treat them as such.
- VectorIndexerModel.transform(): Convert categorical feature values to corresponding 0-based indices

Design notes:
- This maintains sparsity in vectors by ensuring that categorical feature value 0.0 gets index 0.
- This does not yet support transforming data with new (unknown) categorical feature values.  That can be added later.
- This is necessary for DecisionTree and tree ensembles.

Reviewers: Please check my use of metadata and my unit tests for it; I'm not sure if I covered everything in the tests.

Other notes:
- This also adds a public toMetadata method to AttributeGroup (for simpler construction of metadata).

CC: @mengxr

### CODE REVIEW

**Comment by sryza** [comment]
@jkbradley sorry for the delay in responding here.  Your breakdown of operations makes sense to me.

A stats collector seems like a good idea.  I also wonder if there's some way to hook it in with Hive table statistics so we can avoid a pass over the data, but maybe that should be saved for future.  If you aren't planning to get to this in the near future, but think you'll have bandwidth to review, I'd be happy to work on it.  Otherwise, I'm happy to look over whatever you put up.

**Comment by jkbradley** [comment]
@sryza  Thanks for offering!  That would be great if you have the bandwidth to work on this.  I'd be happy to help review.

One comment: It would be nice to be able to take advantage of FeatureAttributes in the spark.ml package, but that's a WIP right now: [https://github.com/apache/spark/pull/4460]

**Comment by jkbradley** [comment]
I think it's ready now.  I'll add a quick Java unit test soon to make sure getters/setters work correctly.

**Comment by jkbradley** [comment]
I feel like the code could be a bit shorter; I'll think about that more tomorrow and whether we can make working with DataFrames and metadata easier in general.

**Comment by mengxr** [comment]
@jkbradley I have a question about the expected behavior. Say I have a vector column containing 2 features. One is categorical: 0, 1, 2, 3, 4, 5 and the other is continuous but only takes values from 1.0, 2.0, 4.0. Then if I set `maxCategories` to 10, both will be recognized as categorical and the mapping for the second feature may become something like 1.0 -> 0, 2.0 -> 1, 4.0 -> 2. Is it what we expect?

**Comment by jkbradley** [comment]
@mengxr  Yes, that's what we'd expect.  Eventually, we'd want to be able to specify which features to index, either (a) via another parameter specifying specific features to index or (b) via metadata, where we do not index features which already have metadata.

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    val newField = prepOutputField(dataset.schema, map)
+    val newCol = callUDF(transformFunc, new VectorUDT, dataset(map(inputCol)))
+    // For now, just check the first row of inputCol for vector length.
+    val firstRow = dataset.select(map(inputCol)).take(1)
+    if (firstRow.length != 0) {
+      val actualNumFeatures = firstRow(0).getAs[Vector](0).size
+      require(numFeatures == actualNumFeatures, "VectorIndexerModel expected vector of length" +
+        s" $numFeatures but found length $actualNumFeatures")
+    }
+    dataset.withColumn(map(outputCol), newCol.as(map(outputCol), newField.metadata))
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala]: It'd be nice for withColumn to take metadata.  I'll make a JIRA for that.

**Comment by jkbradley** [comment]
I did some minor cleanups, but I don't see any great places to remove code.  I added a Java test suite.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks, and sorry for the long delay!

**Pull request closed without merging**